core/service.go

```diff
@@ -1,244 +1,962 @@
+	}
+
+	dig, err := svc.Digest()
+	if err != nil {
+		stop()
+		return nil, err
 	}
+
+	return &RunningService{
+		Service: svc,
+		Key: ServiceKey{
+			Digest:   dig,
+			ClientID: clientMetadata.ClientID,
+		},
+		Host:  dialHost,
+		Ports: ports,
+		Stop: func(context.Context) error {
+			stop()
+			// HACK(vito): do this async to prevent deadlock (this is called in Detach)
+			go AllServices.Detach(svcCtx, upstream)
+			var errs []error
+			for _, closeListener := range closers {
+				errs = append(errs, closeListener())
+			}
+			return errors.Join(errs...)
+		},
+	}, nil
 }
 
-type nopCloser struct {
-	io.Writer
+func (svc *Service) startReverseTunnel(ctx context.Context, bk *buildkit.Client) (running *RunningService, err error) {
+	dig, err := svc.Digest()
+	if err != nil {
+		return nil, err
+	}
+
+	host, err := svc.Hostname(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	clientMetadata, err := engine.ClientMetadataFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rec := progrock.RecorderFromContext(ctx)
+
+	svcCtx, stop := context.WithCancel(context.Background())
+	svcCtx = engine.ContextWithClientMetadata(svcCtx, clientMetadata)
+	svcCtx = progrock.RecorderToContext(svcCtx, rec)
+
+	fullHost := host + "." + network.ClientDomain(clientMetadata.ClientID)
+
+	tunnel := &c2hTunnel{
+		bk:                 bk,
+		upstreamHost:       svc.HostUpstream,
+		tunnelServiceHost:  fullHost,
+		tunnelServicePorts: svc.HostPorts,
+	}
+
+	checkPorts := []Port{}
+	for _, p := range svc.HostPorts {
+		desc := fmt.Sprintf("tunnel %s %d -> %d", p.Protocol, p.FrontendOrBackendPort(), p.Backend)
+		checkPorts = append(checkPorts, Port{
+			Port:        p.FrontendOrBackendPort(),
+			Protocol:    p.Protocol,
+			Description: &desc,
+		})
+	}
+
+	check := newHealth(bk, fullHost, checkPorts)
+
+	exited := make(chan error, 1)
+	go func() {
+		exited <- tunnel.Tunnel(svcCtx)
+	}()
+
+	checked := make(chan error, 1)
+	go func() {
+		checked <- check.Check(svcCtx)
+	}()
+
+	select {
+	case err := <-checked:
+		if err != nil {
+			stop()
+			return nil, fmt.Errorf("health check errored: %w", err)
+		}
+
+		return &RunningService{
+			Service: svc,
+			Key: ServiceKey{
+				Digest:   dig,
+				ClientID: clientMetadata.ClientID,
+			},
+			Host:  fullHost,
+			Ports: checkPorts,
+			Stop: func(context.Context) error {
+				stop()
+				return nil
+			},
+		}, nil
+	case err := <-exited:
+		stop()
+		return nil, fmt.Errorf("proxy exited: %w", err)
+	}
+}
+
+type RunningService struct {
+	Service *Service
+	Key     ServiceKey
+	Host    string
+	Ports   []Port
+	Stop    func(context.Context) error
+}
+
+type Services struct {
+	starting map[ServiceKey]*sync.WaitGroup
+	running  map[ServiceKey]*RunningService
+	bindings map[ServiceKey]int
+	l        sync.Mutex
+}
+
+type ServiceKey struct {
+	Digest   digest.Digest
+	ClientID string
+}
+
+// AllServices is a pesky global variable storing the state of all running
+// services.
+var AllServices = &Services{
```

Comment:
We could, I started with it this way because nearly everywhere calls `WithServices` so we would need to pass another value through a ton of APIs, so this was the fast & dirty way to get everything working. I'll try cleaning it up anyway and see how it goes, global state isn't exactly the best pattern to set. :sweat_smile:

Reply:
Totally, on a highly-related note, I've in general been meaning to update our codebase to make more heavy use of `*core.Context` to pass stuff around (`bk *buildkit.Client`, `progSockPath`, the existing `CacheMap`s, etc.).

I was gonna see if I could sneak that into the Checks PR if time, but obviously there's a big one-time annoyance/time-sink cost to pay since it touches so many files in the codebase.

Point being that you could update to just pass the `CacheMap` through that if you agree it makes sense, but don't worry about updating everything else to do that too; we can bring everything into alignment separately.

Reply:
Yeah at this point I think a big old "bag of dependencies that tend to be needed by X layer of abstraction" makes more sense than a ton of fiddly arguments.

I have a small hunch calling it `Context` might become confusing if there are scenarios where we want a different `context.Context` used alongside it, but totally agree in spirit, and maybe it's consistent anyway, and I can't think of a better name besides `BagOfDependencies`.

Anyway I converted everything over to non-global state, it wasn't too bad, just waiting on test results now.